### PR TITLE
Treat internal boundaries in p:f:T

### LINKED ISSUE #12665: fullydistributed triangulation made from description incl. manifolds results in disjoint faces across at subdomain borders
Hi everyone,
I might have found a bug in parallel::fullydistributed::Triangulation (pfdt) when I was trying to create a pfdt from a serial triangulation, although it might also be rooted in wrong use of the functions.
My testcase is a circular disk in 2D with a circular inclusion which I read from an .msh file (a similar mesh can be created combining a `hyper_shell` and a `hyper_ball`):
circ_inclusion
Then I do the following:
1) Read in a grid with `read_msh()` into a serial triangulation.
2) In cells touching the boundary `cell->set_all_manifold_ids(100)`
3) Attach a `SphericalManifold` to both the serial triangulation and pfdt (similar as in `tests/fullydistributed_grids/copy_serial_tria_03.cc`)
4) Refine serial triangulation via `refine_global()`
5) Partition serial triangulation via `partition_triangulation`
6) Create description of the serial triangulation via `create_description_from_triangulation`
7) Actually create pfdt via `create_triangulation()`
If I do so, the created pfdt features disjoint cell faces across subdomains, as can be seen here:
circ_inclusion_3_procs
I can also replace 6) and 7) by `copy_triangulation()`, which gives the expected results, so, no disjoint cells, but also has a completely different partitioning.
But am I using above mentioned function(s) in a wrong way?
Any comments, suggestions, etc. are highly welcome!
My MWE plus the mesh file (in dummy .txt file format):
[circ_inclusion.txt](https://github.com/dealii/dealii/files/7006226/circ_inclusion.txt)
[copy_serial_tria_03.txt](https://github.com/dealii/dealii/files/7006227/copy_serial_tria_03.txt)

EDIT:
If one does not attach a manifold at all, 6) and 7) do work as expected and in `tests/fullydistributed_grids/copy_serial_tria_03.cc` all cells do have the same manifold attached to them. So it seems, that a cell does not know, which manifolds are attached on other processors to map the mesh in a conforming way (its assuming FlatManifold?) - which sounds reasonable! I guess this functionality is simply impossible due to the design of pfdt?

### PR DESCRIPTION
**Pull request by peterrum**
Assign internal coarse-grid boundaries `numbers::internal_face_boundary_id` as boundary ids.

@richardschu could you give this a try? 

fixes #12665 (once ready)

### CODE REVIEW

File: source/grid/tria.cc [hunk 1]
```diff
@@ -4512,7 +4512,14 @@ namespace internal
                  triangulation.active_cell_iterators_on_level(level))
               if (cell->refine_flag_set())
                 {
-                  if (cell->at_boundary())
+                  // check if cell is at en EXTERIOR boundary
```
Comment (bangerth) [inline source/grid/tria.cc:4515]:
```suggestion
                  // check if cell is at an EXTERIOR boundary
```

File: source/grid/tria.cc [hunk 2]
```diff
@@ -4512,7 +4512,14 @@ namespace internal
...
+                  bool flag = false;
+
+                  for (const auto &f : cell->face_iterators())
+                    flag |=
+                      (f->boundary_id() != numbers::internal_face_boundary_id);
```
Comment (bangerth) [inline source/grid/tria.cc:4520]:
```suggestion
                      (f->at_boundary());
```

File: source/grid/tria.cc [hunk 3]
```diff
@@ -4512,7 +4512,14 @@ namespace internal
...
+
+                  if (flag && cell->at_boundary())
```
Comment (bangerth) [inline source/grid/tria.cc:4522]: These two conditions are the same, if you look at what `cell->at_boundary()` does.

File: source/grid/tria.cc [hunk 4]
```diff
@@ -4512,7 +4512,14 @@ namespace internal
                  triangulation.active_cell_iterators_on_level(level))
               if (cell->refine_flag_set())
                 {
-                  if (cell->at_boundary())
+                  // check if cell is at en EXTERIOR boundary
+                  bool flag = false;
```
Comment (bangerth) [inline source/grid/tria.cc:4516]: Give this variable a better name.

File: source/grid/tria.cc [hunk 5]
```diff
@@ -11878,6 +11885,25 @@ Triangulation<dim, spacedim>::create_triangulation(
           }
       }
 
+      {
+        auto cell      = this->begin(level);
+        auto cell_info = cell_infos[level].begin();
+        for (; cell_info != cell_infos[level].end(); ++cell_info)
+          {
+            // find cell that has the correct cell
```
Comment (bangerth) [inline source/grid/tria.cc:11893]: ??

**Review by bangerth — COMMENTED** [review]
Haven't looked at the entire patch, but already have questions :-)

**Comment by richardschu** [comment]
> @richardschu could you give this a try?
> 
> fixes #12665 (once ready)

I will let you know asap ... compiling `peterrum:pft_manifolds` now. thank you very much for the help!

**Comment by bangerth** [comment]
It fails a test:
```
1423/4411 Test #1417: serialization/parallel_fullydistributed_construction_data_1.debug ............***Failed   14.08 sec
make[3]: *** [parallel_fullydistributed_construction_data_1.debug/diff] Error 1
make[2]: *** [CMakeFiles/serialization.parallel_fullydistributed_construction_data_1.debug.diff.dir/all] Error 2
make[1]: *** [CMakeFiles/serialization.parallel_fullydistributed_construction_data_1.debug.diff.dir/rule] Error 2
make: *** [serialization.parallel_fullydistributed_construction_data_1.debug.diff] Error 2
Test serialization/parallel_fullydistributed_construction_data_1.debug: DIFF
===============================   OUTPUT BEGIN  ===============================
[ 50%] Generating parallel_fullydistributed_construction_data_1.debug/interrupt_guard.cc
Scanning dependencies of target serialization.parallel_fullydistributed_construction_data_1.debug
[200%] Building CXX object CMakeFiles/serialization.parallel_fullydistributed_construction_data_1.debug.dir/parallel_fullydistributed_construction_data_1.cc.o
Building CXX object CMakeFiles/serialization.parallel_fullydistributed_construction_data_1.debug.dir/parallel_fullydistributed_construction_data_1.debug/interrupt_guard.cc.o
Linking CXX executable parallel_fullydistributed_construction_data_1.debug/parallel_fullydistributed_construction_data_1.debug
Built target serialization.parallel_fullydistributed_construction_data_1.debug
Scanning dependencies of target serialization.parallel_fullydistributed_construction_data_1.debug.diff
Generating parallel_fullydistributed_construction_data_1.debug/output
Generating parallel_fullydistributed_construction_data_1.debug/diff
serialization/parallel_fullydistributed_construction_data_1.debug: BUILD successful.
serialization/parallel_fullydistributed_construction_data_1.debug: RUN successful.
serialization/parallel_fullydistributed_construction_data_1.debug: DIFF failed. ------ Source: /jenkins/workspace/dealii_PR-12669/tests/serialization/parallel_fullydistributed_construction_data_1.output
serialization/parallel_fullydistributed_construction_data_1.debug: DIFF failed. ------ Result: /home/dealii/build/tests/serialization/parallel_fullydistributed_construction_data_1.debug/output
Check /home/dealii/build/tests/serialization/parallel_fullydistributed_construction_data_1.debug/output /jenkins/workspace/dealii_PR-12669/tests/serialization/parallel_fullydistributed_construction_data_1.output
serialization/parallel_fullydistributed_construction_data_1.debug: DIFF failed. ------ Diff:   /home/dealii/build/tests/serialization/parallel_fullydistributed_construction_data_1.debug/failing_diff
serialization/parallel_fullydistributed_construction_data_1.debug: DIFF failed. ------ First 20 lines of numdiff output:
----------------
##2       #:75  <== 1
##2       #:75  ==> 2
@ Absolute error = 1.0000000000e+0, Relative error = 1.0000000000e+0
##2       #:79  <== 4
##2       #:79  ==> 1
@ Absolute error = 3.0000000000e+0, Relative error = 3.0000000000e+0
##2       #:80  <== 0
##2       #:80  ==> 4294967295
@ Absolute error = 4.2949672950e+9, Relative error = Inf
##2       #:81  <== 5
##2       #:81  ==> 4
@ Absolute error = 1.0000000000e+0, Relative error = 2.5000000000e-1
##2       #:82  <== 1
##2       #:82  ==> 0
@ Absolute error = 1.0000000000e+0, Relative error = Inf
##2       #:83  <== 0
##2       #:83  ==> 5
@ Absolute error = 5.0000000000e+0, Relative error = Inf
##2       #:84  <== 0
CMakeFiles/serialization.parallel_fullydistributed_construction_data_1.debug.diff.dir/build.make:64: recipe for target 'parallel_fullydistributed_construction_data_1.debug/diff' failed
CMakeFiles/Makefile2:323: recipe for target 'CMakeFiles/serialization.parallel_fullydistributed_construction_data_1.debug.diff.dir/all' failed
CMakeFiles/Makefile2:330: recipe for target 'CMakeFiles/serialization.parallel_fullydistributed_construction_data_1.debug.diff.dir/rule' failed
Makefile:233: recipe for target 'serialization.parallel_fullydistributed_construction_data_1.debug.diff' failed

```

**Comment by richardschu** [comment]
Good morning!
The fix successfully fixes the issue, but as @bangerth mentioned, it fails a test. Executing my MWE, i get the output `log_np` with p being the number of processors:
[log_n1.txt](https://github.com/dealii/dealii/files/7012050/log_n1.txt)
[log_n2.txt](https://github.com/dealii/dealii/files/7012051/log_n2.txt)
[log_n3.txt](https://github.com/dealii/dealii/files/7012052/log_n3.txt)
[log_n4.txt](https://github.com/dealii/dealii/files/7012053/log_n4.txt)
Visually, the mesh looks fine: (wireframe mode in ParaView, close-up of the new mesh in white, old one in orange - so that loose edge in the orange mesh indicates a 'hole')
closeup_processor_edge_white_new_orange_old

Let me know, when I should test again @peterrum !

File: source/grid/tria.cc (hunk 3 above)
Comment (peterrum) [reply source/grid/tria.cc:4522]: It is not:

https://github.com/dealii/dealii/blob/7a2cec9c87a368a978374001666315c60bc0b003/source/grid/tria_accessor.cc#L2862-L2870

one is checking the boundary ids, the other the index of the neighbors.

**Comment by peterrum** [comment]
@richardschu Thanks for checking! I will convert your program to a test once, we have agreed that this is the way to go.

> but as @bangerth mentioned, it fails a test

That is to be expected, since the definition of `TriangulationDescription::CellData` has slightly changed and the test only tests the serialization.

File: source/grid/tria.cc (hunk 3 above)
Comment (bangerth) [reply source/grid/tria.cc]: Ah, I see. That's a nuisance. This seems wrong in the p::f::T case. We should be marking these faces with `internal_face_boundary_id` and make this function return `false`. Is the former what this patch does?

**Comment by kronbichler** [comment]
Do we still need this one or is it superseded by #12676?

**Pull request closed without merging**